specs/phase1/beacon-chain.md

```diff
@@ -752,298 +417,108 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
-```
-
-##### New default validator for deposits
-
-```python
-def get_validator_from_deposit(state: BeaconState, deposit: Deposit) -> Validator:
-    amount = deposit.data.amount
-    effective_balance = min(amount - amount % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
-    next_custody_secret_to_reveal = get_custody_period_for_validator(
-        ValidatorIndex(len(state.validators)),
-        get_current_epoch(state),
+    pending_header.votes = bitwise_or(
+        pending_header.votes,
+        attestation.aggregation_bits
     )
 
-    return Validator(
-        pubkey=deposit.data.pubkey,
-        withdrawal_credentials=deposit.data.withdrawal_credentials,
-        activation_eligibility_epoch=FAR_FUTURE_EPOCH,
-        activation_epoch=FAR_FUTURE_EPOCH,
-        exit_epoch=FAR_FUTURE_EPOCH,
-        withdrawable_epoch=FAR_FUTURE_EPOCH,
-        effective_balance=effective_balance,
-        next_custody_secret_to_reveal=next_custody_secret_to_reveal,
-        all_custody_secrets_revealed_epoch=FAR_FUTURE_EPOCH,
+    # Check if the PendingShardHeader is eligible for expedited confirmation
+    # Requirement 1: nothing else confirmed
+    all_candidates = [
+        c for c in pending_headers if
+        (c.slot, c.shard) == (pending_header.slot, pending_header.shard)
+    ]
+    if True not in [c.confirmed for c in all_candidates]:
+        # Requirement 2: >= 2/3 of balance attesting
+        participants = get_attesting_indices(state, attestationg.data, pending_commitment.votes)
+        participants_balance = get_total_balance(state, participants)
+        full_committee = get_beacon_committee(state, attestationg.data.slot, attestationg.data.shard)
+        full_committee_balance = get_total_balance(state, full_committee)
+        if participants_balance * 3 > full_committee_balance * 2:
+            pending_header.confirmed = True
+```
+
+#### `process_shard_header`
+
+```python
+def process_shard_header(state: BeaconState,
+                         signed_header: Signed[ShardDataHeader]) -> None:
+    header = signed_header.message
+    header_root = hash_tree_root(header)
+    # Verify signature
+    signer_index = get_shard_proposer_index(state, header.slot, header.shard)
+    assert bls.Verify(
+        state.validators[signer_index].pubkey,
+        compute_signing_root(header, get_domain(state, DOMAIN_SHARD_HEADER)),
+        signed_header.signature
+    )
+    # Verify the length by verifying the degree.
+    if header.commitment.length == 0:
+        assert header.degree_proof == G1_SETUP[0]
+    assert (
+        bls.Pairing(header.degree_proof, G2_SETUP[0]) ==
+        bls.Pairing(header.commitment.point, G2_SETUP[-header.commitment.length]))
     )
+    # Get the correct pending header list
+    if compute_epoch_at_slot(header.slot) == get_current_epoch(state):
+        pending_headers = state.current_epoch_pending_shard_headers
+    else:
+        pending_headers = state.previous_epoch_pending_shard_headers
```

Comment:
From https://hackmd.io/@HWeNw8hNRimMm2m2GH56Cw/sharding_proposal:
> The ShardHeader can be included in the beacon chain in that slot or any subsequent slot in the same or next epoch.

The `else` branch should verify that the header's slot is not earlier than the previous epoch. Alternatively, we can have a similar check to the one in `process_attestation':
```
assert data.target.epoch in (get_previous_epoch(state), get_current_epoch(state))
```